### PR TITLE
Require Seed

### PR DESCRIPTION
**Pull request by davidsean**
Fixes #1039 

Description of changes:
 - 
Requires that the user seeds from the script, displays how to do it as part of the runtime error message

PR Checklist
------------
 - [ ] Tests?
   - [ ] Interface
   - [ ] Core 
 - [ ] Docs?

### CODE REVIEW

**Comment by RudolfWeeber** [comment]
The tutorial scripts and corresponding tex files need to be changed, too.

**Comment by hmenke** [comment]
Great!  More global variables.

**Comment by mkuron** [comment]
This is broken on Python 3: https://travis-ci.org/espressomd/espresso/jobs/247486394#L1369

**Comment by jonaslandsgesell** [comment]
This PR seems to have a lot of merge conflicts. Identical seeds may result in synchronization even when the system is setup in  different start configurations. This is due to stochastic thermostats like the langevin thermostat acting as an attractor and the Lyapunov exponent(s) therefore having a negative sign such that trajectories exponentially converge. The Lyapunov exponent is system dependent.

This attracting behaviour has for example been shown in 1D Lennard-Jones systems and I can reproduce this behaviour with espresso.

The remaining question is whether this synchronization may be present in systems which are of interest for us. In this case it is important to seed with different seeds when using multiple simulations to estimate the variance of observables. With same seeds (but different start configurations
) the estimated variance of the observables are wrongly too small due to the correlation of different runs with same seeds of the PRNG in espresso.

Synchronization has for example also been reported in a 3d alanine peptide system:
http://pubs.acs.org/doi/abs/10.1021/ct800573m

Although synchronization might not always happen I would be careful here and rather use a random seed. I would say to be on the safe side we should look at the alanine peptide example and see wether something like this can happen in our systems. If we decide it s not possible we can keep the fixed seeds and maybe just close this PR?

**Comment by kosovan** [comment]
If the user is required to provide a seed, what prevents him from providing a random seed?

**Comment by davidsean** [comment]
@jonaslandsgesell The question is not whether we want a random seed or not. I agree that for production runs a random seed is preferred. There are pros and cons in both directions, for example, the opposite is true in the development stage (to reproduce a crash after an attempted fix).

The issue is: do we want ESPResSo to make that decision silently for the user? The "REQUIRE SEED" trick enables ESPResSo to side-step the issue and to force the user to make a decision. ESPResSo does not know if the user is in a development vs production stage, and should not make that call.

**Comment by davidsean** [comment]
@kosovan 
No problem with the user providing a random seed, the user can also explicitly ask espresso to find its own random seed it if is desired. The point is to force a line somewhere in the script that makes seeding explicit---and necessary.

**Comment by KaiSzuttor** [comment]
i think that here we should have a default behaviour that catches  most of the users. Therefore, i think for most of the users a random seed is safe, while the people who need a fixed seed should really know that they need a fixed seed and can manually fix it. So i think we don’t need this PR.

**Comment by RudolfWeeber** [comment]
> i think that here we should have a default behaviour that catches  most of the users. Therefore, i think for most of the users a random seed is safe, while the people who need a fixed seed should really know that they need a fixed seed and can manually fix it. So i think we don’t need this PR.
We had this discussion several times and the decision made at the
Espresso meeting was to make seeding explicit. 
system.seed=1
system.seed=np.random.random()
Let's stick with it.

**Comment by KaiSzuttor** [comment]
you are right rudolf, still this is my opinion. if we stick to the decision (what we should do), then we should make the seed parameter a System constructor argument, just as it has been done for the box length.

**Comment by jonaslandsgesell** [comment]
+1 for the constructor. Probably the cleanest solution for this behaviour

**Comment by davidsean** [comment]
+1 for the constructor, will thus not require those extra global variables I added.

**Comment by RudolfWeeber** [comment]
On Sat, Feb 03, 2018 at 09:02:43PM +0000, David Sean wrote:
> +1 for the constructor, will thus not require those extra global variables I added.
I'm not sure about the constructor. In constrast to the box
geometry which is always needed when you have particles, quite a few simulations may not require random numbers.

**Comment by KaiSzuttor** [comment]
So how do you want to only make the seed required for these scenarios? Handle them explicitely and hope not to forget one?

**Comment by davidsean** [comment]
sounds like we are back to` if (!user_has_seeded) throw Random::unseeded_error;` in places where random numbers generated.

File: doc/sphinx/advanced_methods.rst [hunk 1]
```diff
@@ -270,54 +270,21 @@ current implementation with the ``COLLISION_DETECTION`` feature.
...
+
 Please contact the Biofluid Simulation and Modeling Group at the
 University of Bayreuth if you plan to use this feature.
 
-This section describes an alternative way to include soft elastic
-objects somewhat different from the previous chapter. In the Immersed
 Boundary Method (IBM), soft particles are considered as an infinitely
 thin shell filled with liquid (see e.g. :cite:`Peskin2002,Crowl2010,KruegerThesis`). When the
 shell is deformed by an external flow it responds by elastic restoring
+>>>>>>> 86ca06ad32a3dfa71547de702e9933fd6b7f6037
```
Comment (hmenke) [inline doc/sphinx/advanced_methods.rst]: Oops!

**Review by hmenke — CHANGES_REQUESTED** [review]

**Comment by RudolfWeeber** [comment]
@fweik, we need to merge this before the QA day. Is there sth left to do, here?

File: testsuite/constraint_shape_based_cylinder.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+from __future__ import division, print_function
```
Comment (KaiSzuttor) [inline testsuite/constraint_shape_based_cylinder.py]: shape based constraints are tested in ```constraint_shape_based.py```. This and the ellipsoid test file should be removed.

File: testsuite/constraint_shape_based_ellipsoid.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
+from __future__ import division, print_function
```
Comment (KaiSzuttor) [inline testsuite/constraint_shape_based_ellipsoid.py]: remove this file

File: testsuite/constraint_shape_based_wall.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
+from __future__ import division, print_function
```
Comment (KaiSzuttor) [inline testsuite/constraint_shape_based_wall.py]: remove this file

File: testsuite/lb_test.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
+from __future__ import print_function
```
Comment (KaiSzuttor) [inline testsuite/lb_test.py]: remove this file

File: testsuite/lb_test_gpu.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
+from __future__ import print_function
```
Comment (KaiSzuttor) [inline testsuite/lb_test_gpu.py]: remove this file

**Review by KaiSzuttor — CHANGES_REQUESTED** [review]

**Comment by KaiSzuttor** [comment]
there might be some more merge regressions to fix...

**Comment by RudolfWeeber** [comment]
github and CI seem to be happy. Any specific thing @KaiSzuttor?

File: doc/tutorials/01-lennard_jones/scripts/lj_tutorial.py [hunk 7]
```diff
@@ -65,7 +65,9 @@
 
 # System setup
 #############################################################
-system              = espressomd.System(box_l=[1.0, 1.0, 1.0])
+system              = espressomd.Systembox_l=([1.0, 1.0, 1.0])
```
Comment (fweik) [inline doc/tutorials/01-lennard_jones/scripts/lj_tutorial.py]: This looks wrong.

**Review by fweik — CHANGES_REQUESTED** [review]

File: src/core/random.hpp [hunk 8]
```diff
@@ -83,6 +89,12 @@ void init_random_seed(int seed);
 
 inline double d_random() {
   using namespace Random;
+
+  if (!user_has_seeded && !unseeded_error_thrown) {
+    unseeded_error_thrown = true;
```
Comment (KaiSzuttor) [inline src/core/random.hpp]: there is a lot of code duplication here for every function that returns random numbers... also there is no need to throw if temperature is 0.0, e.g. the user only wants the damping part of a langevin thermostat

**Review by KaiSzuttor — CHANGES_REQUESTED** [review]
@davidsean please remove code duplication and only throw if temperature != 0.0

File: testsuite/npt.py [hunk 9]
```diff
@@ -73,7 +74,7 @@ def test_npt(self):
         print(avp,compressibility)
 
         self.assertAlmostEqual(2.0, avp, delta=0.02)
-        self.assertAlmostEqual(0.2, compressibility, delta=0.01)
+        self.assertAlmostEqual(0.2, compressibility, delta=0.013)
```
Comment (KaiSzuttor) [inline testsuite/npt.py:77]: Why did you change the accepted deviation?

**Comment by RudolfWeeber** [comment]
On Tue, Jul 24, 2018 at 06:29:24AM -0700, Kai Szuttor wrote:
> KaiSzuttor commented on this pull request.
> 
> 
> 
> > @@ -73,7 +74,7 @@ def test_npt(self):
>          print(avp,compressibility)
>  
>          self.assertAlmostEqual(2.0, avp, delta=0.02)
> -        self.assertAlmostEqual(0.2, compressibility, delta=0.01)
> +        self.assertAlmostEqual(0.2, compressibility, delta=0.013)
> 
> Why did you change the accepted deviation?
Blew up for a certain number of cores.

**Comment by KaiSzuttor** [comment]
I broke the drude testcase, no idea how. Thermalized bond and harmonic bond tests pass.

**Comment by KaiSzuttor** [comment]
@fweik , ready for review i think

**Review by KaiSzuttor — APPROVED** [review]

File: src/core/random.hpp [hunk 10]
```diff
@@ -1,40 +1,58 @@
...
 #include <string>
 #include <vector>
+#include <stdexcept>
 
 namespace Random {
 extern std::mt19937 generator;
 extern std::normal_distribution<double> normal_distribution;
 extern std::uniform_real_distribution<double> uniform_real_distribution;
+extern bool user_has_seeded;
+extern bool unseeded_error_thrown;
```
Comment (fweik) [inline src/core/random.hpp]: This should not be global, I think the best solution is to make it a static local variable in check_user_has_seeded.

**Review by fweik — CHANGES_REQUESTED** [review]

File: src/core/random.hpp (hunk 10 above)
Comment (KaiSzuttor) [reply src/core/random.hpp]: hmm, but how to switch the user_has_seeded variable?

**Comment by jonaslandsgesell** [comment]
What was the exact reason for not requiring a user provided seed when initializing a random number generator (and preventing a random number generator to exist without a user provided seed or state)? That would be the correct dependency inversion right? Then we would not need to check for a seed to be present all the time.

File: src/core/random.hpp (hunk 10 above)
Comment (fweik) [reply src/core/random.hpp]: I'm only talking about `unseeded_error_thrown`.

**Comment by KaiSzuttor** [comment]
good idea, but i think that would be some more work. having our manpower in mind, i think this PR serves our current needs. But you can open an issue.

**Comment by KaiSzuttor** [comment]
@fweik
